File: konane-ai/konane/KonaneUtils.ts
export const BLACK: Player = "black";
export const WHITE: Player = "white";
export const BLACK_CHECKER = "X";
export const WHITE_CHECKER = "O";
export const EMPTY = ".";
const REMOVE = "remove";
const MOVE = "move";

export type Player = "white" | "black";
export type Empty = ".";
export type Checker = "X" | "O";
export type Cell = Checker | Empty;

export const cellIsChecker = (cell: Cell): cell is Checker => {
  return cell === "X" || cell === "O";
};

export type RemoveChecker = {
  player: Player;
  type: "remove";
  cell: [number, number];
};

export type MoveChecker = {
  player: Player;
  type: "move";
  from: [number, number];
  to: [number, number];
};

export type Action = RemoveChecker | MoveChecker;

export type Jump = {
  offset: [0, 2] | [0, -2] | [2, 0] | [-2, 0];
  times: 1 | 2 | 3;
};

export const possibleJumps: Jump[] = [
  // right
  { offset: [0, 2], times: 1 },
  { offset: [0, 2], times: 2 },
  { offset: [0, 2], times: 3 },
  // left
  { offset: [0, -2], times: 1 },
  { offset: [0, -2], times: 2 },
  { offset: [0, -2], times: 3 },
  // up
  { offset: [2, 0], times: 1 },
  { offset: [2, 0], times: 2 },
  { offset: [2, 0], times: 3 },
  // down
  { offset: [-2, 0], times: 1 },
  { offset: [-2, 0], times: 2 },
  { offset: [-2, 0], times: 3 },
];

type Tuple<TItem, TLength extends number> = [TItem, ...TItem[]] & {
  length: TLength;
};

export type Board<T> = Tuple<T, 8>;

export const actionIsMoveChecker = (action: Action): action is MoveChecker => {
  const { type } = action;
  return type === "move";
};

export const actionIsRemoveChecker = (
  action: Action
): action is RemoveChecker => {
  const { type } = action;
  return type === "remove";
};

export type KonaneGameState = {
  action: Action;
  board: Cell[][];
  turn: number;
  playerValues: {
    white: number;
    black: number;
  };
};

export type ComputerDifficulty = 0 | 1 | 2 | 3 | 4 | 5;

export const konaneDifficulties: { [key: string]: ComputerDifficulty } = {
  novice: 0,
  easy: 1,
  medium: 2,
  hard: 3,
  challenger: 4,
  grandmaster: 5,
};

export const stringIsPlayer = (s: string): s is Player => {
  return s === WHITE || s === BLACK;
};

export const verboseCellPosition = (pos: [number, number]) => {
  const [row, col] = pos;
  return [col + 1, 8 - row];
};

File: konane-ai/konane/KonaneUtils.kt
package konane

const val BLACK = "BLACK"
const val WHITE = "WHITE"
const val BLACK_CHECKER = "X";
const val WHITE_CHECKER = "O";
const val EMPTY = ".";

typealias Cell = Pair<Int, Int> 

typealias PlayerActions = Map<Cell, List<Action>>

typealias Jump = Pair<Cell, Int>

typealias Board = List<MutableList<String>>

fun Board.verbose(): String {
	var s = ""
	for (i in this.indices) {
		s += this[i].toString()
		if (i != this.size - 1) s += "\n"
	}
	return s
}

fun Board.deepCopy(): Board {
	val copy = this.map {
		row -> row.map {
			el -> el
		}.toMutableList()
	}
	return copy
}

open class Action(val player: String, val type: String)

class RemoveChecker(player: String, type: String, val cell: Cell): Action(player, type) {
	override fun toString(): String{
		return "${player} ${type} ${cell}"
	}
}

class MoveChecker(player: String, type: String, val from: Cell, val to: Cell): Action(player, type){
	override fun toString(): String{
		return "${player} ${type} ${from} ${to}"
	}
}

val possibleJumps = listOf(
	// right
	Jump(Cell(0,2), 1),
	Jump(Cell(0,2), 2),
	Jump(Cell(0,2), 3),
	// left
	Jump(Cell(0,-2), 1),
	Jump(Cell(0,-2), 2),
	Jump(Cell(0,-2), 3),
	// up
	Jump(Cell(2,0), 1),
	Jump(Cell(2,0), 2),
	Jump(Cell(2,0), 3),
	// down
	Jump(Cell(-2,0), 1),
	Jump(Cell(-2,0), 2),
	Jump(Cell(-2,0), 3),
)

fun oppositeColor(player: String): String {
	if (player != BLACK && player != WHITE) return ""
	if (player == BLACK) return WHITE
	return BLACK
}

fun randInt(min: Int, max: Int): Int {
	if (max < min) return 0
	return (min..max).random()
}

fun <T>randElement(l: List<T>): T? {
	if (l.size == 0) return null
	val randIdx = randInt(0, l.size - 1)
	return l[randIdx]
}